jws/jws-mediator/src/main/java/org/wso2/financial/services/apim/mediation/policies/jws/JwsRequestHeaderValidationMediator.java

```diff
@@ -0,0 +1,596 @@
+
+import java.io.IOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+import java.text.ParseException;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import javax.naming.InvalidNameException;
+import javax.naming.ldap.LdapName;
+import javax.naming.ldap.Rdn;
+import javax.ws.rs.HttpMethod;
+
+/**
+ * Mediator for validating JWS request headers.
+ */
+public class JwsRequestHeaderValidationMediator extends AbstractMediator {
+
+    private static final Log log = LogFactory.getLog(JwsRequestHeaderValidationMediator.class);
+
+    private String jwSignatureHeaderName;
+    private String validTrustAnchor = null;
+    private String jwsSupportedAlgorithms = null;
+
+    /**
+     * JWK retrieved from JwksURI matching to the KeyID.
+     */
+    private JWK matchingJWK = null;
+
+    @Override
+    public boolean mediate(MessageContext messageContext) {
+
+        org.apache.axis2.context.MessageContext axis2MessageContext = ((Axis2MessageContext) messageContext)
+                .getAxis2MessageContext();
+        Map<String, String> headers = (Map<String, String>)
+                axis2MessageContext.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+
+        String payload = JwsUtils.extractPayload(axis2MessageContext, headers);
+        String jwSignature = headers.get(jwSignatureHeaderName);
+
+        if (StringUtils.isEmpty(jwSignature)) {
+            String errorDescription = "Empty JWS Signature";
+            log.error(errorDescription);
+            setErrorResponseProperties(messageContext, "Bad Request", errorDescription, "400");
+            throw new SynapseException(errorDescription);
+        }
+
+        JSONObject applicationDataJson = new JSONObject((String) messageContext
+                .getProperty(JwsConstants.APPLICATION_DATA_PROPERTY));
+
+        JSONArray additionalSpProperties = JwsUtils.getAdditionalSpProperties(applicationDataJson);
+        String jwksUrl = JwsUtils.getAdditionalSpProperty(additionalSpProperties, JwsConstants.JWS_SP_PROPERTY_KEY);
+
+        /*
+        Check if the payload is blank or matches an empty string ("").
+        If the REST method is POST or PUT, it indicates that the payload is required for these methods.
+         */
+        if (StringUtils.isBlank(payload) || payload.matches("\"\"")) {
+            String restMethod = (String) messageContext.getProperty("REST_METHOD");
+            if (HttpMethod.POST.equals(restMethod) || HttpMethod.PUT.equals(restMethod)) {
+                String errorDescription = "Request payload cannot be empty";
```

Comment:
shall we include the request coming http verb and url also for log message

Reply:
Fixed with 55d7e83f22cd8b61c36a2ec94553099c99069c10